### ISSUE TITLE
Handling/ignoring of the EINTR when reading from socket.

### ISSUE THREAD

**Issue by golkarn** [andymccurdy/redis-py#565]
I am using redis to implement a blocking queue.  In the implementation, I added a signal handler to handle graceful shutdown. 

```
#!/bin/env python
import signal
import redis


CLI = redis.StrictRedis(host="localhost", port=6379)
STOP = False


def signal_handler(sig_num, frame):
    global STOP

    print "stop requested."
    CLI.rpush("my_queue", "sentinel")  # force blpop to return
    STOP = True


signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)


def main():
    item = None

    print "starting..."

    for item in CLI.blpop('my_queue'):
        if STOP:
            print "cleanup and stopped."
            return

        print "processing %s." % (item,)


if __name__ == "__main__":
    main()
```

After upgrading to redis-py v2.10.0, I can no longer send a TERM signal and have it shutdown gracefully.

git bisect pointed to this commit which introduced the issue:
https://github.com/andymccurdy/redis-py/commit/97c50ccebf1498212c8f8e1fa9b87c3b10c48666#diff-b3be2d0fa4e0cf1f1c403f2de43ecdb4

PythonParser (in v2.9.1) uses socket fileobject to read data -- which ignores the EINTR exception.  PythonParser in v2.10.0 uses the SocketBuffer to read data which does not ignore the EINTR exception.  As a result, an ConnectionError gets thrown, and client.execute_command() will retry the command -- which blocks.

Note: The HiredisParser doesn't handle EINTR, so it always had this issue.

What should be the correct behavior when the EINTR exception is raised?

**Comment by thomasst** [comment]
Redis-py should retry EINTR, see https://github.com/andymccurdy/redis-py/pull/743

**Comment by andymccurdy** [comment]
This was fixed in #743
